### PR TITLE
Common subexpression elimination

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by chriseth** [comment]
This first iteration breaks up the code at every opcode that does not only modify the stack (i.e. sstore, mstore, jump, call, ...).
It generates an expression DAG from the opcodes, unifying nodes with the same value and folding constants.
Once this DAG is present, it re-generates the code to compute the values which should reside on the stack at the end.

File: libevmcore/CommonSubexpressionEliminator.h [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+{
+
+class AssemblyItem;
+using AssemblyItems = std::vector<AssemblyItem>;
+
+using EquivalenceClassId = unsigned;
+using EquivalenceClassIds = std::vector<EquivalenceClassId>;
+
+/**
+ * Optimizer step that performs common subexpression elimination and stack reorginasation,
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.h]: reorginasation -> reorganisation

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+	return _out;
+}
+
+void CommonSubexpressionEliminator::feedItem(AssemblyItem const& _item)
+{
+	if (_item.type() != Operation)
+	{
+		if (_item.deposit() != 1)
+			BOOST_THROW_EXCEPTION(InvalidDeposit());
+		setStackElement(++m_stackHeight, getClass(_item, {}));
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: `{}` may fail for some compilers

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+				m_stackHeight - 1 - int(instruction) + int(Instruction::SWAP1)
+			);
+		else if (instruction != Instruction::POP)
+		{
+			vector<EquivalenceClassId> arguments(info.args);
+			for (int i = 0; i < info.args; ++i)
+				arguments[i] = getStackElement(m_stackHeight - i);
+			setStackElement(m_stackHeight + info.ret - info.args, getClass(_item, arguments));
+		}
+		m_stackHeight += info.ret - info.args;
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: `_item.deposit()` ?

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+		if (equal(args.begin(), args.end(), m_equivalenceClasses.at(c).second.begin()))
+			return c;
+	}
+	// constant folding
+	if (_item.type() == Operation && args.size() == 2 && all_of(
+				args.begin(),
+				args.end(),
+				[this](EquivalenceClassId eqc) { return m_equivalenceClasses.at(eqc).first->match(Push); }))
+	{
+		auto signextend = [](u256 a, u256 b) -> u256
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: `const& _a` and b

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+		return it->first.second + 1;
+	else
+		return 0;
+}
+
+bool SemanticInformation::breaksBasicBlock(AssemblyItem const& _item)
+{
+	switch (_item.type())
+	{
+	case UndefinedItem:
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: default case: return true;

File: libevmcore/CommonSubexpressionEliminator.h [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+ * the equivalence classes.
+ */
+class CSECodeGenerator
+{
+public:
+	/// @returns the assembly items generated from the given requirements
+	/// @param _currentStack current contents of the stack (up to stack height of zero)
+	/// @param _targetStackContents final contents of the stack, by stack height relative to initial
+	/// @param _equivalenceClasses equivalence classes as expressions of how to compute them
+	AssemblyItems generateCode(
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.h]: document that it should not be called twice

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+
+	// remove surplus elements
+	while (removeStackTopIfPossible())
+	{
+		// no-op
+	}
+
+	// check validity
+	int finalHeight = 0;
+	if (!_targetStackContents.empty())
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: explain what happens here :)

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+			m_classPositions[_c] != c_invalidPosition,
+			OptimizerException,
+			"Element already removed but still needed."
+		);
+		return m_classPositions[_c];
+	}
+	EquivalenceClassIds const& arguments = m_equivalenceClasses.at(_c).second;
+	for (EquivalenceClassId arg: boost::adaptors::reverse(arguments))
+		generateClassElement(arg);
+
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: RFC

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+			"Opcodes with more than two arguments not implemented yet."
+		);
+	for (size_t i = 0; i < arguments.size(); ++i)
+		assertThrow(m_stack[m_stackHeight - i] == arguments[i], OptimizerException, "Expected arguments not present." );
+
+	AssemblyItem const& item = *m_equivalenceClasses.at(_c).first;
+	while (SemanticInformation::isCommutativeOperation(item) &&
+			!m_generatedItems.empty() &&
+			m_generatedItems.back() == AssemblyItem(Instruction::SWAP1))
+		appendSwap(m_stackHeight - 1);
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: RFC - WTF?

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+}
+
+void CSECodeGenerator::appendSwap(int _fromPosition)
+{
+	if (_fromPosition == m_stackHeight)
+		return;
+	int nr = m_stackHeight - _fromPosition;
+	assertThrow(1 <= nr && nr <= 16, OptimizerException, "Stack too deep.");
+	m_generatedItems.push_back(AssemblyItem(swapInstruction(nr)));
+	// only update if they are the "canonical" positions
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: mention copies here

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 11]
```diff
@@ -0,0 +1,529 @@
...
+void CSECodeGenerator::appendDup(int _fromPosition)
+{
+	int nr = 1 + m_stackHeight - _fromPosition;
+	assertThrow(1 <= nr && nr <= 16, OptimizerException, "Stack too deep.");
+	m_generatedItems.push_back(AssemblyItem(dupInstruction(nr)));
+	m_stackHeight++;
+	m_stack[m_stackHeight] = m_stack[_fromPosition];
+}
+
+void CSECodeGenerator::appendSwap(int _fromPosition)
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: maybe add a hint of removal in the name of the function

**Pull request merged**